client/my-sites/earn/memberships/index.jsx

```diff
@@ -301,6 +286,62 @@ class MembershipsSection extends Component {
 		);
 	}
 
+	getWording( subscriber ) {
+		const isOneTime = subscriber?.plan?.renew_interval === 'one-time';
+		const subscriber_email = subscriber?.user.user_email ?? '';
+		const wording = {
```

Comment:
Every time we call the getWording method we are building the wording object which is performing more translations than necessary.

Maybe we could lazyload these translations on a per-call basis:
```
wording = {
     donation: () => {
          button: ...
          confirmation_subheading: ...
          .
          .
          .
          success: ...
     },
     purchase: () => {
          .
          .
          .
     }
}
return wording[ productType ]();
```

We could also aim for a more efficient approach that stores the translated object as an instance property.

Another version would be a mix of both, we can store the object as an instance property and only initialize the type we access at least once.

Reply:
I think these can memoized too: https://reactjs.org/docs/hooks-reference.html#usememo

Reply:
Another alternative is to add a 2nd parameter to the function to indicate the message we're interested in. That way messages would be translated only when needed, as done previously:

```
getWording( subscriber, messageKey ) {
  // ...
  if ( messageKey === 'button' ) {
    if ( isOneTime ) {
      return translate( 'Remove' );
    } 

    return productType === 'donation' ? translate( 'Cancel Donation' ) : translate( 'Cancel Subscription' );
  }
  // ...
}
```

Reply:
Not sure if we're wandering into the bounds of the theoretical rather than the practical - getting a translation is extremely cheap. This function makes 12 calls to translate, [the underlying library apparently is capable of 76 million calls in a second](https://github.com/aduth/tannin#benchmarks).

Consider that we're making http requests to fetch gravatars for each subscriber....

If you're sure it's worthwhile, I'm happy to make some changes though.

Reply:
The current implementation is O(n) * O(m) where n is the number of re-renders of the component and m is the number of translations, there exists a solution that, after initialization, is O(1).

I think efficiency should be baked into everything we do, and the current code can be more efficient.

What would be the reason for not implementing a better solution? Should we settle with the solution that has a subpar algorithmic complexity because the library is capable of 76M calls? The consequences of everyone thinking this way would be that our product would become slower with every single commit.

Reply:
It's not a blocker from my side. I might be picky, but I don't like shipping changes that negatively affect the performance (even if it's just by a super tiny factor that makes it practically imperceptible). At the very least I always try to keep the current performance.

Reply:
> What would be the reason for not implementing a better solution? 

 * To spend the time on higher priority work than optimising a page that the react profiler says renders in 1ms.
 * To avoid increasing the memory usage that some of the optimisations above would incur

Reply:
This isn't a blocker from my side.

Reply:
While changing the wording I took the opportunity to immediately return rather than translate for all cases and then return. Additionally there are fewer calls to translate now anyway. There are now 4 calls to translate per getWording call, rather than 12.

Given half of the messages are per-subscription, approaches using memoization would have to either store per-subscription translations or re-run the sprintf per render. I don't think the tiny bit of extra memory use and (subjectively) small increase in complexity are worthwhile.

The messageKey suggestion wouldn't make much difference - one of the callers would have to call translate three times instead of once, and perhaps would (subjectively) make the code a tiny bit harder to read.

Thank-you for caring  ❤️